### ISSUE TITLE
register_dataframe_accessor causes "Untyped class decorator obscures type of class; ignoring decorator"

### ISSUE THREAD

**Issue by davetapley** [pandas-dev/pandas-stubs#906]
**Describe the bug**
Example from [the docs](https://pandas.pydata.org/pandas-docs/stable/development/extending.html#registering-custom-accessors) causes:
```
Untyped class decorator obscures type of class; ignoring decoratorPylance[reportUntypedClassDecorator](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportUntypedClassDecorator)
```

in VSCode.

**To Reproduce**

In VSCode:

```python
@pd.api.extensions.register_dataframe_accessor("geo")
class GeoAccessor:
    ...
```

**Please complete the following information:**
 - OS: Ubuntu
 - OS Version 20.04
 - python version 3.11
 - version of type checker pyright v1.1.354
 - version of installed `pandas-stubs` v2.2.1.240316

**Additional context**

This fixes it, PR welcome?

```py
from typing import Callable, TypeVar

T = TypeVar('T')  # return type

def register_dataframe_accessor(name: str) -> Callable[[T], T]: ...
```

See also:
- https://github.com/pandas-dev/pandas/issues/58244

**Comment by twoertwein** [comment]
> This fixes it, PR welcome?

Yes - probably should use a more specific TypeVar, maybe `DataFrameT = TypeVar("DataFrameT")`.

**Comment by twoertwein** [comment]
I opened https://github.com/pandas-dev/pandas/pull/58339 for pandas. Feel free to open a similar PR here
